crates/bevy_ecs/hecs/src/entity_builder.rs

```diff
@@ -65,24 +65,40 @@ impl EntityBuilder {
 
     /// Add `component` to the entity
     pub fn add<T: Component>(&mut self, component: T) -> &mut Self {
-        if !self.id_set.insert(TypeId::of::<T>()) {
+        self.add_dynamic(TypeInfo::of::<T>(), unsafe {
+            &*slice_from_raw_parts(
+                &component as *const T as *const u8,
+                std::mem::size_of::<T>(),
+            )
+        });
+        std::mem::forget(component);
+        self
+    }
+
+    /// Add a dynamic component to the entity
+    ///
+    /// You must provide the type info and a slice containing the component data
+    pub fn add_dynamic(&mut self, type_info: TypeInfo, data: &[u8]) -> &mut Self {
+        debug_assert_eq!(type_info.layout().size(), data.len());
```

Comment:
This may need to be `assert_eq` instead of `debug_assert_eq` to guarantee memory safety.

Reply:
We'll add consistency checks in the next iteration.